### ISSUE TITLE
PolyML hierarchical savestates take exponential time to load.

### ISSUE THREAD

**Issue by barakeel** [polyml/polyml#139]
When using hierarchical savestates to save the HOL4 state before each theorem, I found out that loading a child with on parent is about a few milliseconds with 50 parents it takes about 1second and for 100 parents it takes forever.

My current workaround is to only move down the hierarchy (increase the savestate argument) once every 50 theorems which adds small memory costs but make it reasonable to load savestates on theories with at most 50*50 theorems.
As a non-expert doing this, I am not sure which savestates depends on which parent savestates.

**Comment by barakeel** [comment]
For reference, I am using polyml 5.7.1.

**Comment by dcjm** [comment]
I haven't been able to reproduce this with simple examples with Poly/ML alone.  Could you produce an example which shows this and with the lastest version, 5.8.1, and which doesn't require HOL4?

**Comment by barakeel** [comment]
I was not able to replicate the exponential slow down I experienced before. (either in 5.7.1 or 5.8.1). I guess the linear/quadratic overhead was too much of an overhead for me and some got carried away with my first impression.

For example in the listTheory, 
it take 14.2s to load a 200 level deep savestates 
but it takes 70s to load a savestate that is 400 level deep.

I attached a Poly/ML file that does not relies on HOL4 that basically replicates the result in 5.8.1 [savestate.txt](https://github.com/polyml/polyml/files/5103974/savestate.txt)
